File: DSAlgoPython/DSA/SortingAlgos/RadixSort.py
# non comparison algorithm
def radix_sort(lst):
    # Find the maximum element to determine the number of digits
    max_value = max(lst)
    num_digits = len(str(max_value))

    # Perform counting sort for each digit from least significant to most significant
    for digit_index in range(num_digits):
        lst = counting_sort_by_digit(lst, digit_index)

    return lst


def counting_sort_by_digit(lst, digit_index):
    count = [0] * 10  # Counting array for digits 0-9
    output = [0] * len(lst)  # Output array to store sorted elements

    # Count the occurrences of each digit
    for num in lst:
        digit = (num // (10 ** digit_index)) % 10
        count[digit] += 1

    # Calculate the cumulative sum of the count list
    for i in range(1, 10):
        count[i] += count[i - 1]

    # Place elements in the output array based on the digit and original order
    for i in range(len(lst) - 1, -1, -1):
        num = lst[i]
        digit = (num // (10 ** digit_index)) % 10
        output[count[digit] - 1] = num
        count[digit] -= 1

    return output


lst = [170, 45, 75, 90, 802, 24, 2, 66]
sorted_lst = radix_sort(lst)
print(sorted_lst)

File: DataStructures/src/sortingTechniques/RadixSort.java
package sortingTechniques;

import java.util.Arrays;
import java.util.Scanner;

public class RadixSort {

	void countingSort(int array[], int size, int place) {
		int[] output = new int[size + 1];
		int max = array[0];
		for (int i = 1; i < size; i++) {
			if (array[i] > max)
				max = array[i];
		}
		int[] count = new int[max + 1];

		for (int i = 0; i < max; ++i)
			count[i] = 0;

		// Calculate count of elements
		for (int i = 0; i < size; i++)
			count[(array[i] / place) % 10]++;

		// Calculate cumulative count
		for (int i = 1; i < 10; i++)
			count[i] += count[i - 1];

		// Place the elements in sorted order
		for (int i = size - 1; i >= 0; i--) {
			output[count[(array[i] / place) % 10] - 1] = array[i];
			count[(array[i] / place) % 10]--;
		}

		for (int i = 0; i < size; i++)
			array[i] = output[i];
	}

	int getMax(int array[], int n) {
		int max = array[0];
		for (int i = 1; i < n; i++)
			if (array[i] > max)
				max = array[i];
		return max;
	}

	void radixSort(int array[], int size) {
		int max = getMax(array, size);

		for (int place = 1; max / place > 0; place *= 10)
			countingSort(array, size, place);
	}

	public static void main(String args[]) {
		Scanner scn = new Scanner(System.in);
		System.out.println("Enter the elemnts to be sorted. ");

		int n = scn.nextInt();
		int[] data = new int[n];

		System.out.println("Enter the elemnts to be sorted.");
		for (int i = 0; i < n; i++) {
			data[i] = scn.nextInt();
		}

		int size = data.length;
		RadixSort rs = new RadixSort();
		rs.radixSort(data, size);
		System.out.println("Sorted Array in Ascending Order: ");
		System.out.println(Arrays.toString(data));

		scn.close();
	}
}